lib/pbench/server/api/resources/query_apis/datasets/datasets.py

```diff
@@ -0,0 +1,347 @@
+                "params": elastic_options,
+            },
+        }
+
+        if action == "update":
+            painless = "ctx._source.authorization=params.authorization"
+            script_params = {"authorization": {"access": access, "owner": owner}}
+            script = {"source": painless, "lang": "painless", "params": script_params}
+            json["path"] = f"{indices}/_update_by_query"
+            json["kwargs"]["json"]["script"] = script
+        elif action == "get":
+            json["path"] = f"{indices}/_search"
+        elif action == "delete":
+            json["path"] = f"{indices}/_delete_by_query"
+        else:
+            raise APIInternalError(f"requested action {action!r} is unknown")
+
+        current_app.logger.info("{} assembled {}", dataset.name, json)
+        return json
+
+    def postprocess(self, es_json: JSONOBJECT, context: ApiContext) -> Response:
+        """
+        Returns a summary of the returned Elasticsearch query results, showing
+        the list of dictionaries with user selected fields from request json as keys
+        Note: id field is added by server by default whereas other fields are client-selected.
+
+        [
+            {
+                "id": "1c25e9f5b5dfc1ffb732931bf3899878",
+                "@timestamp": "2021-07-12T22:44:19.562354",
+                "run": {
+                    "controller": "dhcp31-171.example.com",
+                    "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19",
+                },
+                "@metadata": {
+                    "controller_dir": "dhcp31-171.example.com"
+                }
+            },
+        ]
+        """
+        # If there are no matches for the user, query, and time range,
+        # return the empty list rather than failing.
+        action = context["action"]
+        dataset = context["dataset"]
+        failures = 0
+        if action == "get":
+            try:
+                count = es_json["hits"]["total"]["value"]
+                if int(count) == 0:
+                    current_app.logger.info("No data returned by Elasticsearch")
+                    return jsonify([])
+            except KeyError as e:
+                raise PostprocessError(
+                    HTTPStatus.BAD_REQUEST,
+                    f"Can't find Elasticsearch match data {e} in {es_json!r}",
+                )
+            except ValueError as e:
+                raise PostprocessError(
+                    HTTPStatus.BAD_REQUEST,
+                    f"Elasticsearch hit count {count!r} value: {e}",
+                )
+            results = []
+            for hit in es_json["hits"]["hits"]:
+                s = hit["_source"]
+                s["id"] = hit["_id"]
+                results.append(s)
+        else:
+            if es_json:
+                fields = ("deleted", "updated", "total", "version_conflicts")
+                results = {f: es_json[f] if f in es_json else None for f in fields}
```

Comment:
Don't want to omit the fields, I don't think, because that's the client return. What to do when {Open|Elastic}search fails to return expected fields is a question I didn't attempt to answer. This seemed an easy workaround ... but maybe I should just fail with an internal backend service error if any of them are missing. 😦 

Why, in the moment, I chose `None`, instead of 0? Yeah, that's a more interesting question, and one I'm incapable of answering convincingly. 😆 This does treat the operation as a "partial failure" and will skip updating or deleting the `Dataset`, which *may* be appropriate since we have an unexpected Elasticsearch response... but if I'm going to worry about that, an internal error directly because of the unexpected response might be more appropriate than just returning a results dict to the caller with `None` "counts". Maybe. 👀 

I'll definitely do something here ...

Reply:
Why I didn't use `get` here, I have no idea. Hmm.